### PR TITLE
NullPointerException trying to retrieve things without ids-parameter

### PR DESCRIPTION
**Pull request by danielFesenmeyer**
A NullPointerException occured when a user tried to retrieve things with an empty ids-parameter, e.g. by running the following REST request:
`GET '/api/2/things?ids=`

The user got status code 503 as a consequence. This PR fixes this by returning status code 400 with an appropriate error message instead.

In this context, we also changed the way missing parameters are handled (for ditto in general). If the user did not specify the ids-parameter at all, she got a 404 status code (akka default for _MissingQueryParamRejection_). We changed this to a 400 response, because we consider such a request as bad input, so "Bad Request" is more helpful than "Not Found".

### CODE REVIEW

File: services/gateway/endpoints/src/main/java/org/eclipse/ditto/services/gateway/endpoints/utils/DittoRejectionHandlerFactory.java [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+import akka.http.javadsl.server.MissingQueryParamRejection;
+import akka.http.javadsl.server.RejectionHandler;
+import akka.http.javadsl.server.Route;
+
+/**
+ * Factory to create a custom {@link akka.http.javadsl.server.RejectionHandler}.
+ */
+public final class DittoRejectionHandlerFactory {
+
+    private static final String MISSING_QUERY_PARAM_TEMPLATE = "Request is missing required query parameter '{0}'";
```
Comment (thjaeckle) [inline services/gateway/endpoints/src/main/java/org/eclipse/ditto/services/gateway/endpoints/utils/DittoRejectionHandlerFactory.java]: I think that this won't work in a MessageFormat. AFAIR when you want a `'` in a string formatted by MessageFormat, you have to write it like this: `''{0}''`

File: signals/commands/things/src/main/java/org/eclipse/ditto/signals/commands/things/exceptions/MissingThingIdsException.java [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+import org.eclipse.ditto.model.base.exceptions.DittoRuntimeException;
+import org.eclipse.ditto.model.base.exceptions.DittoRuntimeExceptionBuilder;
+import org.eclipse.ditto.model.base.headers.DittoHeaders;
+import org.eclipse.ditto.model.things.ThingException;
+
+/**
+ * This exception is thrown when multiple things are requested without specifying IDs.
+ */
+@Immutable
+public class MissingThingIdsException extends DittoRuntimeException implements ThingException {
```
Comment (thjaeckle) [inline signals/commands/things/src/main/java/org/eclipse/ditto/signals/commands/things/exceptions/MissingThingIdsException.java]: There is no generated `serialVersionUid` present in the RuntimeException. Probably just cosmetics, but RuntimeException is Java `Serializable`

File: services/gateway/endpoints/src/main/java/org/eclipse/ditto/services/gateway/endpoints/utils/DittoRejectionHandlerFactory.java (hunk 1 above)
Comment (danielFesenmeyer) [reply services/gateway/endpoints/src/main/java/org/eclipse/ditto/services/gateway/endpoints/utils/DittoRejectionHandlerFactory.java]: Thanks for the finding. I fixed this with commit 293b7a107e76c7684b4d19fd662ff35400fe8822.

File: signals/commands/things/src/main/java/org/eclipse/ditto/signals/commands/things/exceptions/MissingThingIdsException.java (hunk 2 above)
Comment (danielFesenmeyer) [reply signals/commands/things/src/main/java/org/eclipse/ditto/signals/commands/things/exceptions/MissingThingIdsException.java]: Thanks for the hint, fixed with commit 6c7f5318025f12948974c92729e9a0ed18a0809b.

**Pull request merged**